source/module_hsolver/hsolver_lip.cpp

```diff
@@ -0,0 +1,260 @@
+#include "hsolver_lip.h"
+
+#include "diago_bpcg.h"
+#include "diago_cg.h"
+#include "diago_dav_subspace.h"
+#include "diago_david.h"
+#include "module_base/global_variable.h"
+#include "module_base/parallel_global.h" // for MPI
+#include "module_base/timer.h"
+#include "module_base/tool_quit.h"
+#include "module_elecstate/elecstate_pw.h"
+#include "module_hamilt_pw/hamilt_pwdft/global.h"
+#include "module_hamilt_pw/hamilt_pwdft/hamilt_pw.h"
+#include "module_hamilt_pw/hamilt_pwdft/wavefunc.h"
+#include "module_hsolver/diagh.h"
+#include "module_hsolver/diago_iter_assist.h"
+
+#include <algorithm>
+#ifdef USE_PAW
+#include "module_cell/module_paw/paw_cell.h"
+#endif
+namespace hsolver
+{
+
+    template <typename T>
+    HSolverLIP<T>::HSolverLIP(ModulePW::PW_Basis_K* wfc_basis_in)
+    {
+        this->classname = "HSolverLIP";
+        this->wfc_basis = wfc_basis_in;
+        this->diag_ethr = GlobalV::PW_DIAG_THR;
+    }
+
+
+    /*
+        lcao_in_pw
+    */
+    template <typename T>
+    void HSolverLIP<T>::solve(hamilt::Hamilt<T>* pHamilt, // ESolver_KS_PW::p_hamilt
+        psi::Psi<T>& psi,           // ESolver_KS_PW::kspw_psi
+        elecstate::ElecState* pes,          // ESolver_KS_PW::pes
+        psi::Psi<T>& transform,
+        const bool skip_charge)
+    {
+        ModuleBase::TITLE("HSolverLIP", "solve");
+        ModuleBase::timer::tick("HSolverLIP", "solve");
+        std::vector<Real> eigenvalues(pes->ekb.nr * pes->ekb.nc, 0);
+        for (int ik = 0; ik < this->wfc_basis->nks; ++ik)
+        {
+            /// update H(k) for each k point
+            pHamilt->updateHk(ik);
+#ifdef USE_PAW
```

Comment:
there are two important things need either you or me to do (or someone else, if we are lucky):
1. PAW is a important feature that can reduce the computational cost, but LCAO does not support PAW yet due to the complex formulation and implementation, but now it is indeed implemented (at least part of) in PW code, so LCAO_IN_PW can take advantages from both PAW and LCAO, this is very important.
2. have to say present PAW implementation makes the code too long. Someone should work on it to make code more easy to maintain.

Reply:
will the conversation continue?

Reply:
> will the conversation continue?

seems not. LCAO_IN_PW in principle can take advantage from both LCAO and PW in diagonalization and hpsi operation, while the HSolver, ESolver and basis are still strongly coupled presently, which means it is hard to avoid making the program structure in mess is difficult at least for now, but it is indeed an attractive alternative of both PW and LCAO in some cases.